### PR TITLE
Fix overloaded method interop calls when parameter types are not specified

### LINKED ISSUE #41493: [Bug]: Incorrect error message when parameter types are not specified for interoperably called overloaded methods
### Description

$Subject

### Steps to Reproduce

In java class `Foo.java`
```java
public static BString testOverloadedMethods(BString str1, long l, BString str2) {
    return str1;
}

public static BString testOverloadedMethods(int i, double d, BString str) {
    return str;
}
```

```ballerina
import ballerina/jballerina.java;

public function testOverloadedMethods(string s1, int i, string s2) returns string = @java:Method {
    'class:"app/java/Foo"
} external;
```

Output:
```
{ballerina/jballerina.java}METHOD_SIGNATURE_DOES_NOT_MATCH 'Incompatible param type for method 'testOverloadedMethods' in class 'app.java.Foo': Java type 'int' will not be matched to ballerina type 'string''
```

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #41493

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 1]
```diff
@@ -251,28 +250,57 @@ private JMethod resolve(JMethodRequest jMethodRequest, List<JMethod> jMethods) {
...
+    private boolean isOverridden(Method method1, Method method2, Class<?> clazz) {
+        // Static methods cannot be overridden
+        if ((Modifier.isStatic(method1.getModifiers()) ^ Modifier.isStatic(method2.getModifiers())) ||
+                method1.getParameterCount() != method2.getParameterCount()) {
+            // This occurs when there are static methods and instance methods and the static method has one more
+            // parameter than the instance method. Also, this occurs when an interop method in an object maps to
+            // instance methods of which one accepting self and another that doesn't.
+            throw new JInteropException(
+                    OVERLOADED_METHODS, "Overloaded methods cannot be differentiated. Please specify the " +
+                    "parameterTypes for each parameter in 'paramTypes' field in the annotation");
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:278]:
```suggestion
                    "parameter types for each parameter in the 'paramTypes' field in the annotation");
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 2]
```diff
@@ -251,28 +250,57 @@ private JMethod resolve(JMethodRequest jMethodRequest, List<JMethod> jMethods) {
...
+        Method currentMethod;
+        Method otherMethod;
+        if (method2.getReturnType().isAssignableFrom(method1.getReturnType())) {
+            currentMethod = method1;
+            otherMethod = method2;
+        } else if (method1.getReturnType().isAssignableFrom(method2.getReturnType())) {
+            currentMethod = method2;
+            otherMethod = method1;
+        } else {
+            return false;
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:290]: Shall we add tests for the not-covered cases?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 3]
```diff
@@ -252,28 +250,57 @@ private JMethod resolve(JMethodRequest jMethodRequest, List<JMethod> jMethods) {
...
-                        // This occurs when there are static methods and instance methods and the static method
-                        // has one more parameter than the instance method. Also this occurs when an interop
-                        // method in an object maps to instance methods of which one accepting self and another
-                        // that doesn't.
-                        throw new JInteropException(
-                                OVERLOADED_METHODS, "Overloaded methods cannot be differentiated. Please specify the " +
-                                "parameterTypes for each parameter in 'paramTypes' field in the annotation");
-                    }
-                    return Optional.of(ithMethod);
+                Method ithMethod = (Method) jMethods.get(i).getMethod();
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:259]: We can move this to outer loop

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 4]
```diff
@@ -252,28 +250,57 @@ private JMethod resolve(JMethodRequest jMethodRequest, List<JMethod> jMethods) {
...
+                Method kthMethod = (Method) jMethods.get(k).getMethod();
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:260]: Shall we directly pass jMethods.get(k).getMethod() to overidden methid instead to use variables

Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:260]:
```suggestion
                Method kthMethod = (Method) jMethods.get(k).getMethod();
```
```suggestion
                Method method2 = (Method) jMethods.get(k).getMethod();
```

File: tests/jballerina-unit-test/src/test/resources/test-src/javainterop/overloading/overridden_method_test.bal [hunk 5]
```diff
@@ -9,6 +10,10 @@ function newCar(handle strName, handle strModel) returns handle = @java:Construc
     'class:"org.ballerinalang.test.javainterop.overloading.pkg.Car"
 } external;
 
+function newSportsCar(handle strName, handle strModel, int seatCount) returns handle = @java:Constructor {
+    'class:"org.ballerinalang.test.javainterop.overloading.pkg.SportsCar"
+} external;
+
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/resources/test-src/javainterop/overloading/overridden_method_test.bal:16]:
```suggestion
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 6]
```diff
@@ -519,7 +524,7 @@ private boolean isValidParamBType(Class<?> jType, BType bType, boolean isLastPar
                 case TypeTags.MAP:
                 case TypeTags.RECORD:
                     return this.classLoader.loadClass(BMap.class.getCanonicalName()).isAssignableFrom(jType);
-                case TypeTags.JSON:
+                case TypeTags.JSON, TypeTags.READONLY:
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:527]:
``` suggestion
                case TypeTags.JSON:
                case TypeTags.READONLY:
```
or else we can have `,` approach for other cases too

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/overloading/pkg/SportsCar.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com) All Rights Reserved.
```
Comment (HindujaB) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/overloading/pkg/SportsCar.java:2]:
``` suggestion
 * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com).
```

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Pull request merged**